apps/prairielearn/assets/scripts/instructorAssessmentAccessClient.ts

```diff
@@ -0,0 +1,225 @@
+import { Temporal } from '@js-temporal/polyfill';
+import { on } from 'delegated-events';
+import morphdom from 'morphdom';
+
+import { onDocumentReady, decodeData } from '@prairielearn/browser-utils';
+import { formatDate } from '@prairielearn/formatter';
+import { html } from '@prairielearn/html';
+
+import { Modal } from '../../src/components/Modal.html.js';
+
+import { AccessRulesTable } from './lib/accessRulesTable.js';
+import { EditAccessRuleModal } from './lib/editAccessRuleModal.js';
+
+onDocumentReady(() => {
+  const enableEditButton = document.getElementById('enableEditButton');
+  const editModeButtons = document.getElementById('editModeButtons');
+  const accessRulesTable = document.querySelector('.js-access-rules-table');
+  const editAccessRuleModal = document.querySelector('.js-edit-access-rule-modal');
+  const deleteAccessRuleModal = document.querySelector('.js-delete-access-rule-modal');
+
+  const accessRulesData = decodeData('access-rules-data');
+
+  const ptHost = (accessRulesTable as HTMLElement)?.dataset.ptHost ?? '';
+  const devMode = (accessRulesTable as HTMLElement)?.dataset.devMode === 'true';
+  const hasCourseInstancePermissionView =
+    (accessRulesTable as HTMLElement)?.dataset.hasCourseInstancePermissionView === 'true';
+  const timezone = (accessRulesTable as HTMLElement)?.dataset.timezone ?? 'UTC';
+
+  let editMode = false;
+
+  function refreshTable() {
+    morphdom(
+      accessRulesTable as Node,
+      AccessRulesTable({
+        accessRules: accessRulesData,
+        ptHost,
+        devMode,
+        hasCourseInstancePermissionView,
+        editMode,
+      }).toString(),
+    );
+  }
+
+  refreshTable();
+
+  const addRuleButton = document.getElementById('addRuleButton');
+
+  enableEditButton?.addEventListener('click', () => {
+    editMode = true;
+    enableEditButton.style.display = 'none';
+    editModeButtons?.style.removeProperty('display');
+    addRuleButton?.style.removeProperty('display');
+    refreshTable();
+  });
+
+  on('click', '#saveAndSyncButton', () => {
+    const form = document.getElementById('accessRulesForm') as HTMLFormElement;
+    const assessmentAccessRulesInput = form.querySelector(
+      'input[name="assessment_access_rules"]',
+    ) as HTMLInputElement;
+    if (!assessmentAccessRulesInput) return;
+    accessRulesData.forEach((accessRule: Record<string, any>) => {
+      accessRule.startDate = adjustedDate(accessRule.start_date).toISOString().slice(0, 19);
+      accessRule.endDate = adjustedDate(accessRule.end_date).toISOString().slice(0, 19);
+      accessRule.timeLimitMin = parseInt(accessRule.time_limit);
```

Comment:
You'll have to be careful here. In particular, this property must be an integer if it's defined:

https://github.com/PrairieLearn/PrairieLearn/blob/fc1f40c7438bb3d28cc98e28c193ca59ad70c751/apps/prairielearn/src/schemas/schemas/infoAssessment.json#L231

However, as currently implemented, if I save an access rule with the time limit input empty, it gets serialized to `null` and then fails to sync.

Reply:
This is now parsed as an integer in [1dfef07](https://github.com/PrairieLearn/PrairieLearn/pull/9934/commits/1dfef07527ed0296ff3cdfa7c58d85fce2304b97). Also, I am now filtering the rules and removing any null or undefined values.